lib/bundler/cli/outdated.rb

```diff
@@ -97,105 +93,81 @@ def run
-              end
-            end
-
-            gems.each do |gem|
-              print_gem(
-                gem[:current_spec],
-                gem[:active_spec],
-                gem[:dependency],
-                groups,
-                options_include_groups.any?
-              )
-            end
-          end
-        else
-          outdated_gems_list.each do |gem|
-            print_gem(
-              gem[:current_spec],
-              gem[:active_spec],
-              gem[:dependency],
-              gem[:groups],
-              options_include_groups.any?
-            )
-          end
-        end
+          Bundler.ui.info(header_group_message(groups)) unless options[:parseable]
 
-        exit 1
+          print_gems(gems)
+        end
+      else
+        print_gems(outdated_gems_list)
       end
+
+      perform_exit
     end
 
   private
 
-    def retrieve_active_spec(strict, definition, current_spec)
-      if strict
-        active_spec = definition.find_resolved_spec(current_spec)
-      else
-        active_specs = definition.find_indexed_specs(current_spec)
-        if !current_spec.version.prerelease? && !options[:pre] && active_specs.size > 1
-          active_specs.delete_if {|b| b.respond_to?(:version) && b.version.prerelease? }
-        end
-        active_spec = active_specs.last
-      end
+    def perform_exit
+      exit(outdated_gems_list.empty? ? 0 : 1)
+    end
 
-      active_spec
+    def header_outdated_message
+      return "Outdated gems included in the bundle:" unless options[:pre]
+      "Outdated gems included in the bundle (including pre-releases):"
     end
 
-    def display_nothing_outdated_message(filter_options_patch)
-      unless options[:parseable]
-        if filter_options_patch.any?
-          display = filter_options_patch.map do |o|
-            o.sub("filter-", "")
-          end.join(" or ")
+    def header_group_message(groups)
+      groups ? "===== Group #{groups} =====" : "===== Without group ====="
+    end
 
-          Bundler.ui.info "No #{display} updates to display.\n"
-        else
-          Bundler.ui.info "Bundle up to date!\n"
-        end
+    def nothing_outdated_message
+      return "Bundle up to date!\n" if filter_options_patch.empty?
+
+      display = filter_options_patch.map do |o|
+        o.sub("filter-", "")
+      end.join(" or ")
+
+      "No #{display} updates to display.\n"
+    end
+
+    def retrieve_active_spec(definition, current_spec)
+      return definition.find_resolved_spec(current_spec) if strict
+
+      active_specs = definition.find_indexed_specs(current_spec)
+      if !current_spec.version.prerelease? && !options[:pre] && active_specs.size > 1
+        active_specs.delete_if {|b| b.respond_to?(:version) && b.version.prerelease? }
+      end
+      active_specs.last
+    end
+
+    def print_gems(gems_list)
+      gems_list.each do |g|
+        print_gem(g[:current_spec], g[:active_spec], g[:dependency], g[:groups])
```

Comment:
I agree with extracting the logic, but I like the previous style, and also the more descriptive variable name (`gem`, instead of `g`). Let's avoid the style changes and only extract the common logic the method, keeping it as:

```ruby
def print_gems(gems_list)
  gems_list.each do |gem|
    print_gem(
      gem[:current_spec],
      gem[:active_spec],
      gem[:dependency],
      gem[:groups]
    )
  end
end
```

Reply:
what do you think in kwargs usage?

Reply:
You mean here? No, let's keep it as it is. For new code it's good, but I don't want to convert all of our code base to use it.